server/mediaserver.go

```diff
@@ -489,12 +310,84 @@ func endRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 		if monitor.Enabled {
 			monitor.LogStreamEndedEvent(cxn.nonce)
 		}
+		cxn.eof <- struct{}{}
 		delete(s.rtmpConnections, mid)
 
 		return nil
 	}
 }
 
+func (s *LivepeerServer) startSession(cxn *rtmpConnection) *BroadcastSession {
+
+	mid := rtmpManifestID(cxn.stream)
+	cpl := cxn.pl
+	var sess *BroadcastSession
+
+	// Connect to the orchestrator. If it fails, retry for as long
+	// as the RTMP stream is alive; maybe the orchestrator hasn't
+	// received the block containing the job yet
+	broadcastFunc := func() error {
+		var err error
+		sess, err = selectOrchestrator(s.LivepeerNode, cpl)
+		if err == ErrDiscovery {
+			return nil // discovery disabled, don't retry
+		} else if err != nil {
+			// Should be logged upstream
+		}
+		s.connectionLock.RLock()
+		defer s.connectionLock.RUnlock()
+		_, active := s.rtmpConnections[mid]
+		if active {
+			return err
+		}
+		return nil // stop if inactive
+	}
+	expb := backoff.NewExponentialBackOff()
+	expb.MaxInterval = BroadcastRetry
+	expb.MaxElapsedTime = 0
+	backoff.Retry(broadcastFunc, expb)
+	return sess
+}
+
+func (s *LivepeerServer) startSessionListener(cxn *rtmpConnection) {
+	mut := sync.Mutex{}
+	cv := sync.NewCond(&mut)
+	pending := false
+	runStartSession := func() {
+		sess := s.startSession(cxn)
+		mut.Lock()
+		defer mut.Unlock()
+		cxn.sess = sess
+		pending = false
+		cv.Signal()
+	}
+	glog.V(common.DEBUG).Info("Starting broadcast listener for ", cxn.pl.ManifestID())
+	for {
+		select {
+
+		case <-cxn.needOrch:
+			go func() {
+				mut.Lock()
+				defer mut.Unlock()
+				if pending {
+					// Skip if we already have a request inflight
+					return
+				}
+				cxn.sess = nil
+				pending = true
+				go runStartSession()
+				for pending {
+					cv.Wait()
```

Comment:
Hmmm I need some help to really understand CondVars  ... why do we need to do cond.Wait() here? From what I'm reading in the docs (https://golang.org/pkg/sync/#Cond.Wait), it unlocks the Mutex and suspends the execution of the goroutine (which in this case starts on line 369). But, aren't we unlocking the Mutex in `runStartSession` too? Why wouldn't this conflict with the Locking happening on line 358?

Reply:
The purpose of a condvar is to block execution of the thread (`wait`). The thread is unblocked once the condvar is `signal`'d. So here, we basically `wait` until the `runStartSession` goroutine completes.

The mutex within the condvar is to ensure atomicity of the wait/signal operations; it prevents things like multiple threads calling `signal` at the same time, interleaving of `wait`/`signal` across multiple threads, etc. (The docs say that holding the lock isn't required to call `signal` but I'm sure the lock is still acquired as part of the implementation.)

I'm not sure what you mean by locking conflicts. The locking in `runStartSession` is not strictly necessary for the condvar itself but is required to protect the assignment to the `cxn` and `pending` objects. 

The sequence is:
```
* Goroutine 1: Acquire lock
* Goroutine 1: Wait, release lock (via condvar)
* Goroutine 2: Acquire lock
* Goroutine 2: Signal
* Goroutine 2: Release lock
* Goroutine 1: Acquire lock (via condvar), resume execution
```

Reply:
Ah I see what you mean! What I meant by "locking conflicts" had to do with mutex being unlocked in `runStartSession` then, unlocked again by line 371. I just found a really good tutorial on condVars and mutexes that makes this all make more sense! : https://computing.llnl.gov/tutorials/pthreads/#ConditionVariables.